### PR TITLE
ARROW-621: [C++] Start IPC benchmark suite for record batches, implement "inline" visitor. Code reorg

### PR DESCRIPTION
**Pull request by wesm**
From the benchmarks, the difference between virtual functions and an "inline" switch statement is very small, but it serves to reduce some boilerplate when many of the visit functions are the same

### CODE REVIEW

File: cpp/src/arrow/loader.cc [hunk 1]
```diff
@@ -131,33 +133,24 @@ class ArrayLoader : public TypeVisitor {
...
-  VISIT_PRIMITIVE(UInt64Type);
-  VISIT_PRIMITIVE(HalfFloatType);
-  VISIT_PRIMITIVE(FloatType);
-  VISIT_PRIMITIVE(DoubleType);
-  VISIT_PRIMITIVE(Date32Type);
-  VISIT_PRIMITIVE(Date64Type);
-  VISIT_PRIMITIVE(TimeType);
-  VISIT_PRIMITIVE(TimestampType);
-
-#undef VISIT_PRIMITIVE
```
Comment (wesm) [inline cpp/src/arrow/loader.cc]: This is what I was looking to eliminate

**Comment by wesm** [comment]
```
$ ./release/ipc-read-write-benchmark 
Run on (8 X 3687.58 MHz CPU s)
2017-03-22 19:40:39
Benchmark                                                  Time           CPU Iterations
----------------------------------------------------------------------------------------
BM_WriteRecordBatch/1/min_time:1.000/real_time            45 us         45 us      31193   21.5106GB/s
BM_WriteRecordBatch/4/min_time:1.000/real_time            48 us         48 us      29289   20.3822GB/s
BM_WriteRecordBatch/16/min_time:1.000/real_time           41 us         41 us      34623   24.0034GB/s
BM_WriteRecordBatch/64/min_time:1.000/real_time           43 us         43 us      32466   22.9193GB/s
BM_WriteRecordBatch/256/min_time:1.000/real_time          67 us         67 us      21216   14.5113GB/s
BM_WriteRecordBatch/1024/min_time:1.000/real_time        169 us        169 us       8204    5.7672GB/s
BM_WriteRecordBatch/4k/min_time:1.000/real_time          566 us        566 us       2420    1.7245GB/s
BM_WriteRecordBatch/8k/min_time:1.000/real_time         1059 us       1059 us       1320   944.317MB/s
BM_ReadRecordBatch/1/min_time:1.000/real_time              1 us          1 us    1893667    1.2952TB/s
BM_ReadRecordBatch/4/min_time:1.000/real_time              2 us          2 us     884226   609.386GB/s
BM_ReadRecordBatch/16/min_time:1.000/real_time             5 us          5 us     277866   185.639GB/s
BM_ReadRecordBatch/64/min_time:1.000/real_time            22 us         22 us      63444   44.7815GB/s
BM_ReadRecordBatch/256/min_time:1.000/real_time           83 us         83 us      17062   11.7476GB/s
BM_ReadRecordBatch/1024/min_time:1.000/real_time         337 us        337 us       4155    2.8965GB/s
BM_ReadRecordBatch/4k/min_time:1.000/real_time          1414 us       1414 us       1010   707.448MB/s
BM_ReadRecordBatch/8k/min_time:1.000/real_time          2863 us       2863 us        486   349.244MB/s
```

**Comment by wesm** [comment]
This is failing because of PARQUET-919, sorry about that. https://github.com/apache/parquet-cpp/pull/272

**Review by xhochy — APPROVED** [review]
+1, LGTM

Eventhough it might not make a large performance difference, it really improves on code readability.

**Pull request closed without merging**